monkey/monkey_island/cc/services/agent_plugin_service/mongo_agent_plugin_repository.py

```diff
@@ -0,0 +1,226 @@
+            if plugin_type not in configuration_schemas:
+                configuration_schemas[plugin_type] = {}
+            configuration_schemas[plugin_type][plugin_name] = config_schema_dict
+
+        return configuration_schemas
+
+    def get_all_plugin_manifests(self) -> Dict[AgentPluginType, Dict[str, AgentPluginManifest]]:
+        manifest_dicts = self._agent_plugins_collection.find(projection=["plugin_manifest"])
+        manifests: Dict[AgentPluginType, Dict[str, AgentPluginManifest]] = {}
+
+        for manifest_dict in manifest_dicts:
+            try:
+                manifest = AgentPluginManifest(**manifest_dict["plugin_manifest"])
+            except Exception as err:
+                raise RetrievalError(
+                    f"Can't create plugin manifest from the data in the database: {err}"
+                ) from err
+            plugin_type = manifest.plugin_type
+            plugin_name = manifest.name
+            if plugin_type not in manifests:
+                manifests[plugin_type] = {}
+            manifests[plugin_type][plugin_name] = manifest
+
+        return manifests
+
+    def store_agent_plugin(self, operating_system: OperatingSystem, agent_plugin: AgentPlugin):
+        plugin_name = agent_plugin.plugin_manifest.name
+        plugin_type = agent_plugin.plugin_manifest.plugin_type
+        try:
+            plugin_dict = self._get_agent_plugin(plugin_type, plugin_name)
+        except Exception:
+            plugin_dict = self._encode_agent_plugin(agent_plugin)
+            plugin_dict[BINARY_OS_MAPPING_KEY] = {}
+
+        if operating_system.value in plugin_dict[BINARY_OS_MAPPING_KEY]:
+            raise StorageError(
+                f"Plugin {plugin_name} of type {plugin_type} already has a binary for "
+                f"operating system {operating_system}"
+            )
+
+        try:
+            id = self._agent_plugins_binaries_collections[operating_system].put(
+                agent_plugin.source_archive
+            )
+            plugin_dict[BINARY_OS_MAPPING_KEY][operating_system.value] = id
+        except Exception:
+            raise StorageError(
+                f"Failed to store binary for plugin {plugin_name} of type {plugin_type} for "
+                f"operating system {operating_system}"
+            )
+
+        print(f"Writing plugin {plugin_dict}")
+        self._agent_plugins_collection.update_one(
+            {"plugin_manifest.name": plugin_name, "plugin_manifest.plugin_type": plugin_type.value},
+            {"$set": plugin_dict},
+            upsert=True,
+        )
+
+    def _encode_agent_plugin(self, agent_plugin: AgentPlugin) -> Dict[str, Any]:
+        dict = agent_plugin.dict(simplify=True, exclude={"source_archive"})
+
+        return self._encode_dict(dict)
+
+    def _encode_dict(self, d: dict):
+        new_dict = {}
+        for k, v in d.items():
+            if isinstance(k, Enum):
+                k = k.value
+            if isinstance(v, Enum):
```

Comment:
I thought pydantic serialization handles this?

Reply:
I ran into some problems, but @VakarisZ was able to remove it with no ill effects